circleci/s3-upload

```diff
@@ -5,32 +5,64 @@
 # - All source dir content is uploaded (recursively).
 # - Example s3 bucket URL format: s3://<bucket>/<etc>/
 # - Requires two env vars to be set, with access to the bucket:
-#       * AWS_ACCESS_KEY_ID
-#       * AWS_SECRET_ACCESS_KEY
+#     * AWS_ACCESS_KEY_ID
+#     * AWS_SECRET_ACCESS_KEY
 # - If not set, import them from another app that uploads to the same bucket https://circleci.com/gh/Clever/[APP_NAME]/edit#env-vars
+# - If `--content-encoding gzip` is specified:
+#     * Files will be individually gzipped.
+#     * File names will not be changed (i.e. they will not be extended with .gz).
+#     * Files will all be tagged with `Content-Encoding: gzip` metadata in S3.
+#     * The original source dir content will be left unchanged.
 #
 # Usage:
 #
-#   s3-upload [SOURCE_DIR] [S3_BUCKET_URL]
+#   s3-upload [--content-encoding gzip] SOURCE_DIR S3_BUCKET_URL
 
 set -e
 
 DIR=$(dirname "$0")
 . $DIR/utils
 
-if [ $# -ne 2 ]; then
-    echo "Incorrect number of arguments given. Expected 2, received $#"
-    echo "Usage: s3-upload [SOURCE_DIR] [S3_BUCKET_URL]"
+USAGE="Usage: s3-upload [--content-encoding gzip] SOURCE_DIR S3_BUCKET_URL"
+WORK_DIR="/tmp/s3-upload/work-dir"
+
+if [[ $# -eq 4 ]]; then
+    if [[ $1 != "--content-encoding" ]]; then
+        echo "$USAGE"
+        exit 1
+    fi
+    if [[ $2 != "gzip" ]]; then
+        echo "Valid values for --content-encoding are: gzip"
+        echo "$USAGE"
+        exit 1
+    fi
+    CONTENT_ENCODING=$2
+    SOURCE_DIR=$3
+    S3_BUCKET_URL=$4
+elif [[ $# -eq 2 ]]; then
+    SOURCE_DIR=$1
+    S3_BUCKET_URL=$2
+else
+    echo "$USAGE"
     exit 1
 fi
 
-# User supplied args
-SOURCE_DIR=$1
-S3_BUCKET_URL=$2
+DIR_TO_UPLOAD="$SOURCE_DIR"
+ADDITIONAL_S3_FLAGS=""
+
+if [[ $CONTENT_ENCODING == "gzip" ]]; then
+    rm -rf $WORK_DIR
+    mkdir -p $WORK_DIR
+    cp -r $SOURCE_DIR $WORK_DIR
+    find $WORK_DIR -type f -exec gzip --best {} \; -exec mv {}.gz {} \;
```

Comment:
Hi @arsalansufi 👋 Thanks for making these changes, so awesome! 🎉 🙇 

Do you know what will happen if there are extra files in the upload directory? Is it safe to apply gzip compression to all files types that we find with this command?

Thanks! 🙂

Reply:
Yes good q! This command will gzip every single file it finds in the specified directory. My thinking here was that it's up to you as a consumer to determine whether it's safe to gzip all the files in a given directory. If not, you can make more targeted calls to gzip+upload specified files. Or you can organize your source files differently.